regression/multi-svc-app/multi_svc_app_test.go

```diff
@@ -0,0 +1,250 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package multi_svc_app_test
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+
+	"github.com/aws/copilot-cli/regression/client"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("regression", func() {
+	var cli *client.CLI
+	var expectedWorkloadResponse = make(map[string]string)
+
+	deployWorkloadSpecs := func() {
```

Comment:
nit: what do you think of extracting this function outside of `Describe` and taking the `cli` and `expectedWorkloadResponse` as params to remove the global vars? (just a nit tho, it's pretty clear in the test)
```go
func deployWorkloadsSpec(cli *client.CLI, wantedResponses map[string]string)
```

Reply:
Would the below be an example?
```go
func deployWorkloadsSpec(cli *client.CLI, wantedResponses map[string]string) {}

var _ = Describe("regression", func() {
       When("Using the old CLI", func() {
              var cli *client.CLI
              var wantedResponses map[string]string
              BeforeEach(func() {
                     cli = fromCLI
                     wantedResponses = map[string]string{}
              })

              When("Deploying workloads", func() {
                     deployWorkloadsSpec(cli, wantedResponses)
              })
       })
})
```

This code snippet probably wouldn't work because ginkgo wouldn't be able to construct the spec tree during its ["Tree Construction Phase"](https://onsi.github.io/ginkgo/#mental-model-how-ginkgo-traverses-the-spec-hierarchy). 

It's possible to confine `cli` and `wantedResponse` into each `When("old")` block and `When("new")` block:
```go
var _ = Describe("regression", func() {

       When("Using the old CLI", func() {
              var cli *client.CLI
              BeforeEach(func() {
                     cli = fromCLI
              })

              When("Deploying workloads", func() {
                     It("should succeed", func() {
                            cli.Run()
                     })
              })
       })

       When("Using the new CLI", func() {
              var cli *client.CLI
              BeforeEach(func() {
                     cli = toCLI
              })

              When("Deploying workloads", func() {
                     It("should succeed", func() {
                            cli.Run()
                     })
              })
       })
})
```
However we wouldn't be able to reuse the specs like we are doing now with `deployWorkloadSpecs`, because:
```go
func deployWorkloadsSpec() {
       It("should succeed", func() {
              cli.Run() // Error: cli variable is not accessible
       }
}

var _ = Describe("regression", func() {
       When("Using the old CLI", func() {
              var cli *client.CLI
              BeforeEach(func() {
                     cli = fromCLI
              })

              When("Deploying workloads", deployWorkloadsSpec)
       })
})

// This doesn't work either.
var _ = Describe("regression", func() {
       When("Using the old CLI", func() {
              var cli *client.CLI = fromCLI // This code would be invoked at "tree phase" at which time `fromCLI` is `nil` yet.
              When("Deploying workloads", deployWorkloadsSpec) 
       })
})
```

Because of ginkgo's "Tree Construction Phase" and "Run Phase", we have several constraints:
1. The value of `cli` and `wantedResponses` should be assigned in `BeforeEach` (or `It`) so that the value assignment actually happens during ginkgo's "Run Phase". This means it's required that their declaration are somewhere accessible by `BeforeEach`.
2. If we want to reuse the specs, then `cli` and `wantedResponses` need to be accessible by `deployWorkloadSpecs` as well, in addition to `BeforeEach`.

`deployWorkloadSpecs` is special in that it's supposed to be invoked during "tree phase", not "run phase". Therefore, it can't take any parameters whose actual value needs to be assigned during "run phase". Given all these, and given that we'd like to order the specs with this hierarchy, I've found that the smallest scope of `cli`, `wantedResponses`, and `deployWorkloadSpecs` are probably inside of `Describe`.

Sorry for the lengthy response 🙇🏼‍♀️ ! The "Tree phase" and "Run phase" gave me some headache when I was writing the test suite, so I thought it'd probably be beneficial for me to share some learnings here :meow_wine

Reply:
oh that's really interesting, thanks for sharing!